wip/stacked-borrows.md

```diff
@@ -120,14 +143,15 @@ impl Tracking {
 
 This method will never return the same value twice.
 We say that a `CallId` is *active* if the call stack contains a stack frame with that ID.
+In the following, we pretend there exists a function `barrier_is_active(id)` that can check this.
 
 **Note**: Miri uses a slightly more complex system with a `HashSet<CallId>` tracking the set of active `CallId`; that is just an optimization to avoid having to scan the call stack all the time.
 
 ### Allocating memory
 
 When allocating memory, we have to initialize the `Stack` associated with the new locations, and we have to choose a `Borrow` (a tag) for the initial pointer to this memory.
 
-For most memory, the stack of each freshly allocated memory location is `Stack { borrows: vec![Shr], frozen_since: None }`, and the initial pointer to that memory has tag `Shr(None)`.
+For most memory, the stack of each freshly allocated memory location is `Stack { borrows: vec![Raw], frozen_since: None }`, and the initial pointer to that memory has tag `Alias(None)`.
```

Comment:
Do we also carry the alignment of that memory ? The section below on dereferencing raw pointers states that `size_of_val` is checked, but it says nothing about checking that the pointer is at an alignment boundary of `T`.

Reply:
Or is that implied as part of "the pointer must be dereferenceable" ?

Reply:
Stacked Borrows does not care about alignment at all, and that's why it is not mentioned.

Reply:
Miri complains both on the creation of an unaligned reference and on the pointer dereference below, so I wonder how / where this is tracked ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=4db5abaec7e73cacba66f65c683e9cac)):

```
use core::{arch::x86_64::__m128, mem::transmute};
#[derive(Copy, Clone)]
#[repr(align(8))]
struct U8(u8);
fn main() {
    unsafe {
        type U = U8;
        let a: [U8; 16] = [U8(0); 16];
        let b: *const __m128 = &a as &[U] as *const [U] as *const U as *const __m128;
        let _c: &__m128 = transmute(b);
        let _d = *b;
    }
}
```

Reply:
The former is the validity invariant, and the latter is the normal rule for a memory access.

This question would be answered by a proper definition of MIR semantics (and, in particular, places), which I will write... eventually.^^  But it is not part of Stacked Borrows and IMO should not be answered here.

Reply:
> But it is not part of Stacked Borrows and IMO should not be answered here.

Fair enough, and thanks!